### ISSUE TITLE
Aptly Web UI

### ISSUE THREAD

**Issue by sdumetz** [aptly-dev/aptly#341]
It would make sense to me to have a web UI available as 3rd party software and I'm ready to start something to at least meet my own needs. Before that, I wanted to know if there was any plans to actually do this in the main software (though I don't think it's the best thing to do).

Core features would be (for my use case) : 
- Authentication. It's easy to make aptly only accept localhost requests, but a web frontend would need to be secured in most cases.
- Package upload
- Snapshot creation / list / publish / restore

My core idea is that it's easy to connect the CI to aptly to add new packages, but I need to keep migrations / rollbacks manuals, and an emergency "upload fix" security.

I'd like to know if any likewise project is already ongoing, and if others would benefit from a web UI ?

**Comment by calderaax** [comment]
i was trying to do something similar with python but haven't done much since i don't have much time, (i'm currently working in other projects) but it is a good idea, i could help in my free time if you need.

**Comment by sdumetz** [comment]
I, thanks, it's good to know others have a similar need.
I have worked on something on my [own repo](https://github.com/sdumetz/aptly-web-ui) that's nearly useable, but I don't have yet any plans of making it final or even distribute it.

I need it at work (non-tech people need to make some small repo operations) so I might come up with something relatively generic in a few weeks.

Don't hesitate to look into it and tell me your ideas on how it should be done.

**Comment by calderaax** [comment]
great!! i'll take a look at it. My case is the same we're implemeting aptly to replace our old scripts to manage our repos and mirrors, but the product we're going to deliver is for non-tech people too. I'll try to make some time to work on it  .. Thanks for the link

**Comment by chnyda** [comment]
Hi @calderaax, we have done something in python using pyqt5. The goal is to be able to manipulate publishes and their snapshots graphically. The result is here : https://github.com/chnyda/python-aptly-gui

**Comment by smira** [comment]
@chnyda I'll get a link added to your repo tomorrow. Thanks!

**Comment by mjb2kmn** [comment]
I've been using something I through together in Ruby with Sinatra, but have been wondering if something more tightly integrated with Go would be better. Maybe even bypassing the API. The lack of a mirror API really hurts the idea of a UI based on the REST API.

**Comment by apachelogger** [comment]
The way to deal with shortcomings in the API is to extend the API I'd say.

**Comment by mjb2kmn** [comment]
I agree, @apachelogger. There was such a pull request but it was not accepted and may need additional work.
https://github.com/smira/aptly/pull/166
Even if an official GUI used the API, I'd still suggest it be written in Go.

**Comment by minecraftchest1** [comment]
> Hi @calderaax, we have done something in python using pyqt5. The goal is to be able to manipulate publishes and their snapshots graphically. The result is here : https://github.com/chnyda/python-aptly-gui

Thanks. Your shell completion was a lifesaver.
